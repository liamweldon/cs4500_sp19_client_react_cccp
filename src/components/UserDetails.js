import React from "react";

const UsersDetails = ({users, user, cancel, createUser, updateUser, deleteUser, selectUser, setUser, redirect}) => (
    <div>
        <div>
            <h3>User Details</h3>
            <select
                value={user.id}
                onChange={(e) => selectUser(e.target.value)}
                className="form-control">
                {users.map(user => {
                    //var selected = user.id === user.id ? true : false;
                    var selected = user.id

                    return (
                        <option
                            value={user.id}
                            key={user.id}
                            defaultValue={selected}
                        >
                            {user.id}
                        </option>
                    );
                })}
            </select>
            <br/><br/><label>ID</label>
            <input onChange={(e) =>
                setUser({
                    user: {
                        id: e.target.value,
                        username: user.username,
                        password: user.password,
                        firstName: user.firstName,
                        lastName: user.lastName
                    }
                })} className="form-control"
                   value={user.id}/>
            <br/><label>Username</label>
            <input onChange={(e) =>
                setUser({
                    user: {
                        id: user.id,
                        username: e.target.value,
                        password: user.password,
                        firstName: user.firstName,
                        lastName: user.lastName
                    }
                })} className="form-control"
                   value={user.username}/>
            <br/><label>Password</label>
            <input onChange={(e) =>
                setUser(
                    {
                        user: {
                            id: user.id,
                            username: user.username,
                            password: e.target.value,
                            firstName: user.firstName,
                            lastName: user.lastName
                        }
                    })} className="form-control"
                   value={user.password}/>
            <br/><label>First Name</label>
            <input onChange={(e) =>
                setUser({
                    user: {
                        id: user.id,
                        username: user.username,
                        password: user.password,
                        firstName: e.target.value,
                        lastName: user.lastName
                    }
                })} className="form-control"
                   value={user.firstName}/>
            <br/><label>Last Name</label>
            <input onChange={(e) =>
                setUser({
                    user: {
                        id: user.id,
                        username: user.username,
                        password: user.password,
                        firstName: user.firstName,
                        lastName: e.target.value
                    }
                })} className="form-control"
                   value={user.lastName}/>
            <br/></div>
        <div>
            <button className="btn btn-primary" onClick={() => {
                createUser().then(window.setTimeout(redirect("/admin/users/"), 500))
                window.location.reload();
            }}>
                Create
            </button>
            &nbsp;
            <button className="btn btn-success" onClick={
                () => {
                    updateUser().then(window.setTimeout(redirect("/admin/users/"), 500))
                    window.location.reload();
                }
            }>Update
            </button>
            &nbsp;
            <button className="btn btn-danger" onClick={
                () => {
                    deleteUser().then(window.setTimeout(redirect("/admin/users/"), 500))
                    window.location.reload();
                }
            }>Delete
            </button>
            &nbsp;
            <button className="btn btn-danger" onClick={
                cancel}>
                Cancel
            </button>
            <br/>
            &nbsp;
        </div>
    </div>
);

export default UsersDetails
