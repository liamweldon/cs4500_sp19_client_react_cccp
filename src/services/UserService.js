import {API_ROOT} from '../api-config';

export default class UserService {
    static instance = null;

    static getInstance() {
        if (UserService.instance === null) {
            UserService.instance = new UserService();
        }
        return this.instance;
    }

    findUserById = (userId) => fetch(`${API_ROOT}/api/users/${userId}`).then((response) => response.json());
    findAllUsers = () => fetch(`${API_ROOT}/api/users`).then((response) => response.json());
    // TODO update urls
    createUser = user =>
        fetch(`${API_ROOT}/api/users`,
            {
                method: 'POST',
                body: JSON.stringify(user),
                headers: {'content-type': 'application/json'}
            })
            .then(response => response.json())

    updateUser = user =>
        fetch(`${API_ROOT}/api/users/${user.id}`,
            {
                method: 'PUT',
                body: JSON.stringify(user),
                headers: {'content-type': 'application/json'}
            })
            .then(response => response.json())

    deleteUser = user =>
        fetch(`${API_ROOT}/api/users/${user.id}`,
            {
                method: 'DELETE',
            })
}
