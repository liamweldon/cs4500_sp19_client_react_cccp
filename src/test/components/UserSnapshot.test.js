import React from 'react'
import TestRenderer from 'react-test-renderer'
import UserDetails from '../../components/UserDetails'
import UserService from '../../services/UserService'
import users from '../mockData/users.mock.json'
import mockUserService from '../services/mocks/MockUserService'

const userService = UserService.getInstance();
beforeEach(() => {
    global.fetch = mockUserService(users);
})

// Provide a new service to insert
const test_new_user = {
    "id": 555,
    "username": "alice4",
    "password": "alice3",
    "firstName": "Alice3",
    "lastName": "Wonderland3"
};

// Provide an initial service to populate the edit fields. Replicate the one in the container
// for consistency.
let mock_initial_user = {
    user: {
        username: '',
        firstName: '',
        lastName: '',
    }
}

// Snapshot testing code
test('Creation renders properly in DOM', () => {

    // Create snapshot of initial services
    return userService.findAllUsers().then(users => {
        let component = TestRenderer
            .create(<UserDetails users={users}
                           user={mock_initial_user}/>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // Create a new service via our mocked ServiceService. In the normal case,
        // the ServiceService would modify the database, and the ServiceContainer
        // would then update its state with the new data and re-call the Services
        // funcional component. Here, we simulate that process by updating our
        // mock data via the mock ServiceService and then recreating our
        // functional component with the modified mock data.
        userService.createUser(test_new_user).then(user => {
            component = TestRenderer
                .create(<UserDetails users={users}
                               user={mock_initial_user}/>);
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        }).catch(function (error) {
        });
    }).catch(function (error) {
    });
});