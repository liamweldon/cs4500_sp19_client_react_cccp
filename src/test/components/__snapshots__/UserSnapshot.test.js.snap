// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creation renders properly in DOM 1`] = `
<div>
  <div>
    <h3>
      User Details
    </h3>
    <select
      className="form-control"
      onChange={[Function]}
    >
      <option
        defaultValue={123}
        value={123}
      >
        123
      </option>
      <option
        defaultValue={234}
        value={234}
      >
        234
      </option>
      <option
        defaultValue={345}
        value={345}
      >
        345
      </option>
    </select>
    <br />
    <br />
    <label>
      ID
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      Username
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      Password
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      First Name
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      Last Name
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Create
    </button>
     
    <button
      className="btn btn-success"
      onClick={[Function]}
    >
      Update
    </button>
     
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete
    </button>
     
    <button
      className="btn btn-danger"
    >
      Cancel
    </button>
    <br />
     
  </div>
</div>
`;

exports[`Creation renders properly in DOM 2`] = `
<div>
  <div>
    <h3>
      User Details
    </h3>
    <select
      className="form-control"
      onChange={[Function]}
    >
      <option
        defaultValue={123}
        value={123}
      >
        123
      </option>
      <option
        defaultValue={234}
        value={234}
      >
        234
      </option>
      <option
        defaultValue={345}
        value={345}
      >
        345
      </option>
      <option
        defaultValue={555}
        value={555}
      >
        555
      </option>
    </select>
    <br />
    <br />
    <label>
      ID
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      Username
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      Password
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      First Name
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
    <label>
      Last Name
    </label>
    <input
      className="form-control"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Create
    </button>
     
    <button
      className="btn btn-success"
      onClick={[Function]}
    >
      Update
    </button>
     
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete
    </button>
     
    <button
      className="btn btn-danger"
    >
      Cancel
    </button>
    <br />
     
  </div>
</div>
`;
